src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py

```diff
@@ -403,83 +498,72 @@ def find_fixed_range(testcase_id, job_type):
                                          log_message)
 
-    try:
-      result = _testcase_reproduces_in_revision(testcase, testcase_file_path,
-                                                job_type, middle_revision)
-    except errors.BadBuildError:
-      # Skip this revision.
-      del revision_list[middle_index]
-      max_index -= 1
-      continue
 
+    result, error = _testcase_reproduces_in_revision(testcase, testcase_file_path,
+                                                job_type, middle_revision)
+    if error is not None:
+      if error.error == uworker_msg_pb2.ErrorType.PROGRESSION_BAD_BUILD:
+        # Skip this revision.
+        del revision_list[middle_index]
+        max_index -= 1
+        continue
+      # only bad build errors are recoverable.
+      return error
+    
     if result.is_crash():
       min_index = middle_index
     else:
       max_index = middle_index
 
+    # TODO: this cannot be done from an untrusted bot.
     _save_current_fixed_range_indices(testcase_id, revision_list[min_index],
                                       revision_list[max_index])
 
   # If we've broken out of the loop, we've exceeded the deadline. Recreate the
   # task to pick up where we left off.
-  testcase = data_handler.get_testcase_by_id(testcase_id)
   error_message = ('Timed out, current range r%d:r%d' %
-                   (revision_list[min_index], revision_list[max_index]))
-  data_handler.update_testcase_comment(testcase, data_types.TaskState.ERROR,
-                                       error_message)
-  tasks.add_task('progression', testcase_id, job_type)
+                   (revision_list[min_index], revision_list[max_index]))  
+  return uworker_io.UworkerOutput(
+          testcase=testcase,
+          error_message=error_message,
+          uworker_input=uworker_input,
+          progression_min_revision=revision_list[min_index],
+          progression_max_revision=revision_list[max_index],
+          error=uworker_msg_pb2.ErrorType.PROGRESSION_TIMEOUT)
 
-
-def utask_preprocess(testcase_id, job_type, uworker_env):
-  return uworker_io.UworkerInput(
-      job_type=job_type,
-      testcase_id=testcase_id,
-      uworker_env=uworker_env,
-  )
+def utask_main(uworker_input):
+  """Execute progression task."""
+  return find_fixed_range(uworker_input)
 
 
 def utask_postprocess(output):
-  del output
-
+  if output.error is not None:
+    # TODO(https://github.com/google/clusterfuzz/issues/3008): Change this when
+    # progression is migrated.
+    all_errors = uworker_handle_errors.get_all_handled_errors()
+    uworker_handle_errors.handle(output.error, all_errors)
+    return
+  
+  if output.custom_binary is True:
+    # Retry once on another bot to confirm our results and in case this bot is in
+    # a bad state which we didn't catch through our usual means.
+    testcase = data_handler.get_testcase_by_id(output.uworker_input.testcase_id)
```

Comment:
Do we modify output.testcase at all in uworker_main? if we do then we probably should use it and `put()` it rather than fetching the stale version.

Reply:
Indeed, I'll fix that.

Reply:
Looking at the initial behaviour, It seems we were refetching the testcase after running `testcase_manager.test_for_crash_with_retries`.
The refetch call was here: https://github.com/google/clusterfuzz/pull/3327/files#diff-e67a9c5a80de5ba8fcd9fabf0d42884679d4920f321be15b70088c85d0ee75bbL86
 Was there a particular reason for refetching often?